VUB-HYDR/Wikimpacts
# Data Gap

This PR fills the data gap between event levels. 

Taken as is from #173 (updated checklist, 2024/11/03):

- [x] 1. Admin_Areas: if the joint set of L2/L3 is larger than the one we get in L1,  and automatically, the extra areas need to fill in L1
- [x] 2. Impact categories: if L1 is NULL, but impact information is found in L2 or L3, there are several cases
   - [x] a. L2 is not NULL, L3 is NULL, sum up the numbers from L2, and fill in L1
   - [x] b. L2 is NULL, and L3 is not NULL
     - [x]  - (i) for the same admin_area, and impact in different locations, sum up the numbers and fill a new record in L2
     - [x] - (ii) next, sum up the numbers from L2 to fill in L1

   - [x] c. L2 and L3 are not NULL 
      - [x] - (i) for the same admin_area, and impact in different locations, sum up the numbers in L3 and compare with the number in L2, if the number is smaller, revise the ''Num_Min" to this number, and if the number is larger, then change the Num_max in L2 (ideally this would not happen, because we require the model get the total in L2)
      - [x] - (ii) if the admin_area in L3 is not in L2, sum up the numbers and create a new record in L2
      - [x] - (iii) Next, sum up the all numbers from L2, and fill in L1 with the range 
   - [x] **Special case:** if L1 is not null but the sum of the impact is smaller than L2, use the aggregated L2 number
- [x] 3. time: normally the time information in L2 and L3 is missing, for any case without a year, just fill the year in L2 and L3 using the same as the L1 


Check #101 for more details.

**Update:** ⚠️ 
To be able to apply the data gap and the conversion and inflation adjustment, it would be easier to handle a de-duplicated version of the dataset. So I'm now modifying the db insertion code to create a de-duplicated copy of the data in parquet -- then we can manipulate that easily before a final insertion in the table. I'll work on this while we in the meantime work out the details for _how to adjust certain cases of l3->l2 areas and imapcts (2)_ and _how to handle the data gap before adjusting for inflation (3)_

--------------------

### PR Description

This PR contains a script to fill the data gap as described above, alongside some utility functions in `Database/scr/normalize_data.py` to facilitate the process.
The PR also contains a copy of the **RAW** full run (in chunks of 25 each) with the geojson objects de-duplicated (in `Database/output/full_run_25_deduplicated` and `full_run_25_deduplicated`. The processed version of this with the data gap filled is found in `Database/output/full_run_25_deduplicated_data_gap`. 

This PR **does not** insert the newly processed data into the database because the inflation adjustment and currency conversion steps should happen first so that **Damage** and **Insured Damage** categories are normalized properly! 

### How to test:
- Run the data gap script: 
   ```shell
   poetry run python3 Database/fill_data_gap.py -i Database/output/full_run_25_deduplicated -o Database/output/<OUTPUT_DIR>
    ```
- Check the logs (`data_gap.log`) to see what the data gap has filled (logs exist for everything except for the time gap, which only fills l2/l3 with a start and end year if it's not missing from l1).  **⚠️ Note that new logs would be written to the same file with a timestamp.** 
- Inspect different impact categories manually by loading the parquet files
- Run the data gap script on the files twice: I found that the filling process creates more instances that need to be filled. I don't have time to investigate why, but it may be safe to run the data gap script on the files several times. Ideally, if things went perfectly, there shouldn't be any "data gap filling" left to do when the script is re-run on the same files it output.
